### PR TITLE
feat: add calendarDate and calendarDateRange

### PR DESCRIPTION
**Pull request by co-sic**

### CODE REVIEW

File: src/CalendarDate.ts [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        throw new Error(
+          `CalendarDate Validation Error: Input year ${year} is not valid. Year must be a number between 0 and 9999.`,
+        );
+      }
+      if (month < 0 || month > 11) {
+        throw new Error(
+          `CalendarDate Validation Error: Input month ${month} is not valid. Month must be a number between 0 and 11.`,
+        );
+      }
+      if (day < 1 || day > 31) {
```
Comment (dimitriSaplatkin) [inline src/CalendarDate.ts:51]: you are catching in line 60 that day can be bigger then max days of month. Why we need a guarding block here? The same we can also do for month number

File: src/CalendarDate.ts [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        throw new Error(
+          `CalendarDate Validation Error: Input day ${day} is not valid. Day must be a number between 1 and 31.`,
+        );
+      }
+      let maxDayOfMonth = daysInMonth[month];
+      if (month === 1 && CalendarDate.isLeapYear(year)) {
+        maxDayOfMonth = 29;
+      }
```
Comment (dimitriSaplatkin) [inline src/CalendarDate.ts:59]:
```suggestion
      const maxDayOfMonth = month === 1 && CalendarDate.isLeapYear(year) ? 29 : daysInMonth[month];
```

File: src/CalendarDate.ts [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    if (split.length === 3) {
+      const year = parseInt(split[0]);
+      const month = parseInt(split[1]);
+      const day = parseInt(split[2]);
+      if (!isNaN(year) && !isNaN(month) && !isNaN(day)) {
+        return new CalendarDate(year, month - 1, day);
+      }
+    }
+    throw new Error(
+      `CalendarDate Validation Error: Input ${isoString.toString()} is not a valid ISO Date String.`,
```
Comment (dimitriSaplatkin) [inline src/CalendarDate.ts:90]:
```suggestion
      `CalendarDate Validation Error: Input ${isoString.toString()} is not a valid ISO Date String YYYY-MM-DD.`,
```

File: src/CalendarDate.ts [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+  /**
+   * subtracts the input date from this date and returns the difference in days
+   */
+  getDifferenceInDays(calendarDate: CalendarDate, absolute?: boolean): number {
+    const days = (this.valueOf() - calendarDate.valueOf()) / DAY_IN_SECONDS;
+    return absolute ? Math.abs(days) : days;
+  }
+
+  isInsideOfDateRange(calendarDateRange: CalendarDateRange): boolean {
```
Comment (dimitriSaplatkin) [inline src/CalendarDate.ts:168]: i would keep calenderdate independet from CalenderDate and move it to CalenderDateRange and pass the date instead

File: src/CalendarDateRange.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
+import { CalendarDate } from './CalendarDate';
+
+export class CalendarDateRange {
+  readonly start: CalendarDate;
+
+  readonly end: CalendarDate;
+
+  constructor(start: CalendarDate, end: CalendarDate, autoArrange?: boolean) {
+    if (!autoArrange && end < start) {
```
Comment (dimitriSaplatkin) [inline src/CalendarDateRange.ts:9]: I would set autoArrange to a boolean if not defined

File: src/CalendarDateRange.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+  toJSON(): string {
+    return this.toString();
+  }
+
+  static parseString(string: string): CalendarDateRange {
+    const split = string.split('/');
+    if (split.length !== 2) {
+      throw new Error(
+        `CalendarDateRange Validation Error: Input ${string.toString()} is not valid.`,
```
Comment (dimitriSaplatkin) [inline src/CalendarDateRange.ts:35]:
```suggestion
        `CalendarDateRange Validation Error: Input ${string.toString()} is not valid. Passed string should have the pattern YYYY-MM-DD/YYYY-MM-DD`,
```

File: src/CalendarDateRange.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  toString(): string {
+    return `${this.start.toString()}/${this.end.toString()}`;
+  }
+
+  toJSON(): string {
+    return this.toString();
+  }
+
+  static parseString(string: string): CalendarDateRange {
+    const split = string.split('/');
```
Comment (dimitriSaplatkin) [inline src/CalendarDateRange.ts:32]: maybe it makes sense to check via regEx. What do you think?

File: src/CalendarDateRange.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+      );
+    }
+    return new CalendarDateRange(new CalendarDate(split[0]), new CalendarDate(split[1]));
+  }
+
+  getDifferenceInDays(): number {
+    return this.end.getDifferenceInDays(this.start);
+  }
+
+  isInsideOfDateRange(calendarDateRange: CalendarDateRange): boolean {
```
Comment (dimitriSaplatkin) [inline src/CalendarDateRange.ts:45]: Can we talk about this function?

**Review by dimitriSaplatkin — CHANGES_REQUESTED** [review]

**Review by kkhanhluu — APPROVED** [review]

File: src/CalendarDateRange.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+  includes(
+    calendarDate: CalendarDate,
+    options?: { excludeStart?: boolean; excludeEnd?: boolean },
+  ): boolean;
+
+  includes(
+    calendarDateRange: CalendarDateRange,
+    options?: { excludeStart?: boolean; excludeEnd?: boolean },
+  ): boolean;
```
Comment (dimitriSaplatkin) [inline src/CalendarDateRange.ts:56]: what is the porpuse of defining this parts?

File: test/helpers.ts [hunk 10]
```diff
@@ -0,0 +1,12 @@
+import { CalendarDate } from 'calendar-date';
+
+export const DAY_IN_SECONDS = 86400;
+const DAYS_IN_MONTH = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
+
+export function ensureValidDay(year: number, month: number, day: number): number {
```
Comment (dimitriSaplatkin) [inline test/helpers.ts:6]: why this function is not tested in own file as an official helper method?

**Review by dimitriSaplatkin — CHANGES_REQUESTED** [review]

**Review by dimitriSaplatkin — DISMISSED** [review]

File: src/CalendarDate.ts [hunk 11]
```diff
@@ -61,11 +61,11 @@ export class CalendarDate {
     Object.freeze(this);
   }
 
-  private static isLeapYear(year: number) {
+  public static isLeapYear(year: number) {
```
Comment (dimitriSaplatkin) [inline src/CalendarDate.ts:64]: public static isLeapYear(year: number): boolean {

**Review by dimitriSaplatkin — CHANGES_REQUESTED** [review]

**Review by dimitriSaplatkin — APPROVED** [review]

**Pull request merged**